extensions/amp-ad-network-base/0.1/amp-ad-network-base.js

```diff
@@ -0,0 +1,204 @@
+    this.expandedAdUrl_ = null;
+
+    /** @private {?ArrayBuffer} */
+    this.unvalidatedBytes_ = null;
+
+    /** @private {!SizeInfoDef} */
+    this.initialSize_ = {
+      width: this.element.getAttribute('width'),
+      height: this.element.getAttribute('height'),
+    };
+  }
+
+  /**
+   * @param {string} adUrl
+   * @protected
+   */
+  bindAdRequestUrl(adUrl) {
+    if (this.adUrl_) {
+      dev().warn(TAG, `Ad Request URL already bound: '${this.adUrl_}'`);
+    }
+    this.adUrl_ = adUrl;
+  }
+
+  /**
+   * @param {ValidationResultType} resultType
+   * @param {function()} renderer
+   * @protected
+   */
+  bindRenderer(resultType, renderer) {
+    if (this.boundRenderers_[resultType]) {
+      dev().warn(TAG, `Rendering mode already bound for type '${resultType}'`);
+    }
+    this.boundRenderers_[resultType] = renderer;
+  }
+
+  /**
+   * @param {function()} validator
+   * @protected
+   */
+  bindValidator(validator) {
+    if (this.boundValidator_) {
+      dev().warn(TAG, 'Validator already bound.');
+    }
+    this.boundValidator_ = validator;
+  }
+
+  /**
+   * Collapses slot by setting its size to 0x0.
+   */
+  forceCollapse() {
+    super.attemptChangeSize(0, 0);
+  }
+
+  /**
+   * @return {string} The finalized ad request URL.
+   * @private
+   */
+  getExpandedUrl_() {
+    dev().assert(this.adUrl_, 'Ad Request URL never registered!');
+    // TODO add expansion logic
+    this.expandedAdUrl_ = this.adUrl_;
+    return this.expandedAdUrl_;
+  }
+
+  /**
+   * @param {string} creative
+   * @return {!RenderingDataInputDef}
+   */
+  getRenderingDataInput_(creative) {
+    return /** @type {!RenderingDataInputDef} */ ({
+      creativeMetadata: getAmpAdMetadataMock(creative, TAG, ''),
+      // TODO(levitzky) This may change based on the ad response.
+      size: this.initialSize_,
+      adUrl: this.expandedAdUrl_ ? this.expandedAdUrl_ : this.adUrl_,
+    });
+  }
+
+  /**
+   * Processes the ad response as soon as the XHR request returns. This can be
+   * overridden and used as a hook to perform any desired logic before passing
+   * the response to the validator.
+   * @param {?Promise<?{bytes: !ArrayBuffer, headers: !Headers}>} response
+   * @protected
+   */
+  handleAdResponse(response) {
+    const unvalidatedBytes = response.bytes;
+    const headers = response.headers;
+    dev().assert(this.boundValidator_, 'Validator never bound!');
+    this.unvalidatedBytes_ = unvalidatedBytes;
+    this.boundValidator_(unvalidatedBytes, headers, this)
+        .then(validatedBytes =>
+          this.handleValidationResponse(utf8Decode(validatedBytes)));
+  }
+
+  handleAdResponseError(error) {
+    // TODO(levitzky) add actual error processing logic.
+    dev().warn(TAG, error);
+  }
+
+  /**
+   * Processes validation response and delegates further action to appropriate
+   * renderer.
+   * @param {string} validatedResponse The utf-8 decoded ad response.
+   */
+  handleValidationResponse(validatedResponse) {
```

Comment:
can you order methods so that public methods go first. and put method caller before callee.